### PR TITLE
issue 3037: Fix android compilation warnings

### LINKED ISSUE #3037: Fix android compilation warnings
### User Story

As a developer, I don't want warnings when running commands.

### Description

*Type*: Bug

Running `react-native run-android` generates a bunch of warnings. We want to remove as much as possible.

Examples of warnings:

```
Configuration 'compile' in project ':app' is deprecated. Use 'implementation' instead.
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)

Configuration 'compile' in project ':react-native-splash-screen' is deprecated. Use 'implementation' instead.
Configuration 'testCompile' in project ':react-native-splash-screen' is deprecated. Use 'testImplementation' instead.
WARNING: The specified Android SDK Build Tools version (25.0.3) is ignored, as it is below the minimum supported version (26.0.2) for Android Gradle Plugin 3.0.1.
Android SDK Build Tools 26.0.2 will be used.
To suppress this warning, remove "buildToolsVersion '25.0.3'" from your build.gradle file, as each version of the Android Gradle Plugin now has a default version of the build tools.

/home/julien/Documents/Projects/status-react/android/app/src/main/AndroidManifest.xml:23:5-87 Warning:
        uses-permission#android.permission.BLUETOOTH was tagged at AndroidManifest.xml:23 to remove other declarations but no other declaration present
/home/julien/Documents/Projects/status-react/android/app/src/main/AndroidManifest.xml:25:5-92 Warning:
        uses-permission#android.permission.WRITE_CONTACTS was tagged at AndroidManifest.xml:25 to remove other declarations but no other declaration present
/home/julien/Documents/Projects/status-react/android/app/src/main/AndroidManifest.xml:28:5-90 Warning:
        uses-permission#android.permission.RECORD_VIDEO was tagged at AndroidManifest.xml:28 to remove other declarations but no other declaration present

> Task :app:hemroidBuild
/home/julien/Android/Sdk/ndk-bundle/build/core/setup-app.mk:81: Android NDK: Application targets deprecated ABI(s): mips64 armeabi mips    
/home/julien/Android/Sdk/ndk-bundle/build/core/setup-app.mk:82: Android NDK: Support for these ABIs will be removed in a future NDK release.  
```

Some might be related to our dependencies: in this case we must create issues in relevant project if needed.

### PR DESCRIPTION
**Pull request by pacamara**
Fixes #3037 "Fix android compilation warnings"

### Summary:

Remove nearly all android build warnings, using sed to modify build.gradle files of node_modules. For node_modules already forked by status there are separate PRs (see refs below).

4 warnings left:
* "Task.leftShift(Closure) method has been deprecated" -- from node_modules/realm/android/build.gradle. Have tested fix for this, but cannot use sed as fix is multiline. Should we fork realm or leave this one?
* "registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)" -- from one of the plugins, not from status code.
* "warning: the transform cache was reset" -- from XCode build, not android.
* "Configuration 'compile' in project ':app' is deprecated. Use 'implementation' instead." -- [stackoverflow](https://stackoverflow.com/questions/47300679/configuration-compile-in-is-deprecated-but-all-configurations-are-implementa?rq=1) says could also be from a plugin.

The warnings (now silenced) about unnecessary removal of BLUETOOTH, WRITE_CONTACTS and RECORD_VIDEO permissions were leftovers from when some node_modules were adding these. 

### Testing notes:
Need to run scripts/setup again before building.

### Steps to test:
- scripts/setup
- build as normal
- observe log for build warnings

status: ready

### CODE REVIEW

File: android/app/build.gradle
Comment (jeluard) [inline android/app/build.gradle]: What is the potential downside here? Looks like there was a reason for this difference in version?

File: android/app/src/main/AndroidManifest.xml
Comment (jeluard) [inline android/app/src/main/AndroidManifest.xml:25]: Are those not necessary anymore?

File: android/app/src/main/AndroidManifest.xml
Comment (jeluard) [inline android/app/src/main/AndroidManifest.xml]: This one is still necessary?

**Review by jeluard — CHANGES_REQUESTED** [review]
Thanks!

I'd rather not have scripts hot patching dependencies, seems brittle.
Would you mind creating issues in the related projects? This also includes the one we already forked as we don't intend to diverge from original upstreams.

File: android/app/src/main/AndroidManifest.xml
Comment (pacamara) [reply android/app/src/main/AndroidManifest.xml:25]: Yes, looks like BLUETOOTH, WRITE_CONTACTS and RECORD_VIDEO were introduced by upstream projects sometime in the past, but these problems have now been fixed.
```
~/dev/status-react/node_modules $ grep -r BLUETOOTH  * | grep AndroidManifest
~/dev/status-react/node_modules $ grep -r RECORD_VIDEO  * | grep AndroidManifest
~/dev/status-react/node_modules $ grep -r WRITE_CONTACTS  * | grep AndroidManifest
react-native-contacts/README.md:Add permissions to your `android/app/src/main/AndroidManifest.xml` file.  Add only the permissions you need (i.e. if you don't need the _WRITE_CONTACTS_ permission then there's no need to add it).
~/dev/status-react/node_modules $ 
```
 So there is no longer a need to remove them from AndroidManifest.xml, and trying to do so gives the warning.

File: android/app/src/main/AndroidManifest.xml
Comment (pacamara) [reply android/app/src/main/AndroidManifest.xml]: Yes, looks like it's pulled in by instabug dependency:
```
~/dev/status-react/node_modules $ grep -r RECORD_AUDIO  * | grep AndroidManifest
instabug-reactnative/android/build/intermediates/exploded-aar/com.instabug.library/instabug/3.4.0/AndroidManifest.xml:    <uses-permission android:name="android.permission.RECORD_AUDIO" />
~/dev/status-react/node_modules $ 
```

**Comment by pacamara** [comment]
> I'd rather not have scripts hot patching dependencies, seems brittle.
> Would you mind creating issues in the related projects?

Sure no problem, will start creating the issues. :smile: 

Note though there is precedent for patching this way, pkg-hacks.js already does this for the javascript files, creating the "hacking..." lines of output during scripts/setup.

File: android/app/build.gradle
Comment (pacamara) [reply android/app/build.gradle]: It's not essential to remove this as 26.0.2 is already the minimum supported version for Android Gradle Plugin 3.0.1. But there's no longer a need to specify it explicitly as the gradle plugin now specifies a default buildToolsVersion. Felt it was better to go with that, what do you think?

For the upstream projects these lines must be removed by updateUnforkedNodeModules.sh else we get errors like
```
WARNING: The specified Android SDK Build Tools version (25.0.3) is ignored, as it is below the minimum supported version (26.0.2) for Android Gradle Plugin 3.0.1.
Android SDK Build Tools 26.0.2 will be used.
To suppress this warning, remove "buildToolsVersion '25.0.3'" from your build.gradle file, as each version of the Android Gradle Plugin now has a default version of the build tools.
```

File: android/app/src/main/AndroidManifest.xml
Comment (jeluard) [reply android/app/src/main/AndroidManifest.xml]: Could you update the comment to add this info?

**Comment by jeluard** [comment]
@pacamara Right this is part of `rn-nodeify` but I'd rather move in the direction of removing those.

It's ok if there are more warning, it's better to address those cleanly in upstream dependencies (or remove those completely) in the long run IMO.

File: android/app/src/main/AndroidManifest.xml
Comment (pacamara) [reply android/app/src/main/AndroidManifest.xml]: Done!

**Comment by pacamara** [comment]
@jeluard Ah I see, got it. Have:
* created all the issues in upstream projects. 
* added the comment about RECORD_AUDIO.
* also changed from `api` to `implementation` for the dependencies directly used by status-react, because `api` is only recommended for the dependencies used by library projects (such as the upstream node_modules).

To be clear, do you want `updateUnforkedNodeModules.sh` removed from this PR?

**Comment by jeluard** [comment]
> To be clear, do you want updateUnforkedNodeModules.sh removed from this PR?

I would say yes as it could hide stuff by mistake. It's ok if we have some warnings due to 3rd party deps.

Thanks!

**Comment by pacamara** [comment]
Roger that :smile: 
Have:
* moved the script to a gist: https://gist.github.com/pacamara/b69fef08920b06fbfdf536c1be355d05
* also updated react-native-svg to 6.0.1-rc.3 in package.json, the maintainer has already fixed it
* that required upgrading android/app/build.gradle to `compileSdkVersion 27`

The build warnings from node_modules are back but do not include react-native-svg.

File: android/app/build.gradle
Comment (jeluard) [inline android/app/build.gradle:119]: Does that impact our target mobile version?

**Review by jeluard — APPROVED** [review]

**Comment by jeluard** [comment]
@pacamara Awesome thanks for the changes!

File: android/app/build.gradle
Comment (pacamara) [reply android/app/build.gradle:119]: No there is no impact on the target version. :smile: 
From https://medium.com/google-developers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd (article by a dev at google):
> It should be emphasized that changing your compileSdkVersion does not change runtime behavior... 
> Therefore it is strongly recommended that you always compile with the latest SDK.

**Comment by pacamara** [comment]
@jeluard You're welcome! :beers:

**Comment by pacamara** [comment]
@jeluard Wondering if we want to hold off from asking upstream maintainers of node_modules to upgrade to gradle 3? It's a breaking change for freshly generated react native apps, unless manual changes are made to the generated code. See
https://github.com/react-native-community/react-native-svg/issues/581

Perhaps we should wait until React Native itself is upgraded to gradle 3... what do you think?

**Comment by jeluard** [comment]
@pacamara Makes sense. Do you know if RN has planned such a change?

**Comment by pacamara** [comment]
> Makes sense. Do you know if RN has planned such a change?

@jeluard Can't find an open issue in react-native relating to a planned upgrade. Issue https://github.com/facebook/react-native/issues/16536 suggests users upgrade gradle manually. Shall I close the issues in upstream node_modules projects? Or could ask maintainers to create a separate gradle 3 tag?

**Comment by jeluard** [comment]
@pacamara Sounds good! I guess we can wait now for more movement from RN. Maintainers are free to close your issues if they feel like it I guess?

**Comment by dmitryn** [comment]
`run-android` works fine for me (except warnings coming from node_modules which were discussed above). 

Building in Android Studio gives me 2 errors though:

```
/Users/trybeee/Library/Android/sdk/ndk-bundle/build/core/setup-app.mk
Error:(81) Android NDK: Application targets deprecated ABI(s): mips64 armeabi mips    
Error:(82) Android NDK: Support for these ABIs will be removed in a future NDK release.    
```

But it doesn't build in `develop` branch either.

**Review by dmitryn — APPROVED** [review]

**Comment by pacamara** [comment]
@dmitryn Pushed another commit to fix the ABI problem. I was not experiencing it because my NDK version was out of date (now upgraded to v16):
* set APP_ABI to only non-deprecated ABIs per https://developer.android.com/ndk/guides/abis.html
* also upgrade APP_PLATFORM to match minSdkVersion per https://developer.android.com/ndk/guides/application_mk.html

> I guess we can wait now for more movement from RN. Maintainers are free to close your issues if they feel like it I guess?

@jeluard Sure! :beers:

**Comment by pacamara** [comment]
Hi @jeluard @dmitryn , it's still showing as pending maintainer confirmation, could you confirm the payout please?

**Comment by dmitryn** [comment]
@pacamara where do you see it? Did you get the tokens to your address?

**Comment by pacamara** [comment]
@dmitryn They just arrived a few minutes ago -- thank you! :beers:  

> where do you see it?

 Previously, the "pending maintainer confirmation" status was on the issue page and also on the SOB page.

**Pull request merged**
